### PR TITLE
Fix max_depth recursion crash in AsynchronousLoader

### PR DESCRIPTION
**Pull request by DaGuT**
Some people use dict samples with strings, numpy array etc. They are returned from dataloader and are passed to model. Later those are parsed/replaced/removed inside of step/forward function.

Example of such dict (I use it):
`{'image': tensor([WHATEVER]), 'path': [WHATEVER_STRING], 'target': tensor([WHATEVER_NUMPY]), 'meta': {'PatientAge': tensor([WHATEVER]), 'PatientSex': tensor([WHATEVER]), 'StudyDate': tensor([WHATEVER])}, 'mask': tensor([WHATEVER]), 'bboxes': [[tensor([WHATEVER]), tensor([WHATEVER]), tensor([WHATEVER])]]}`

Sadly, async loader was crashing as max depth of recursion was reached (e.g. when it meets string, it just infinitely goes to else in original AsyncLoader). 

**I went to torch.utils.data.default_collate (as it's said that code is based on it), took few lines from there and it works now.** :) 

So it will no longer crash at named_tuples, dict-like objects and strings. Those will all be processed.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   (no need, it's just a small bugfix)
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes issue when anything other than torch.tensor containing thing is passed to AsynchronousLoader (more details above)

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by DaGuT** [comment]
Small note:
Idk if it is related to the patch, but with AsyncLoader it works almost twice slower than without.
